governance/status.go

```diff
@@ -0,0 +1,413 @@
+	Arch        string            `json:"arch"`
+	Containers  []ContainerStatus `json:"containerStatus"`
+}
+
+func (w WorkloadStatus) String() string {
+	return fmt.Sprintf("AgreementId: %v, "+
+		"WorkloadURL: %v, "+
+		"Org: %v, "+
+		"Version: %v, "+
+		"Arch: %v, "+
+		"Containers: %v",
+		w.AgreementId, w.WorkloadURL, w.Org, w.Version, w.Arch, w.Containers)
+}
+
+func (w WorkloadStatus) IsSame(compare WorkloadStatus) bool {
+	if w.AgreementId != compare.AgreementId || w.WorkloadURL != compare.WorkloadURL || w.Org != compare.Org || w.Version != compare.Version || w.Arch != compare.Arch {
+		return false
+	} else if len(w.Containers) != len(compare.Containers) {
+		return false
+	} else {
+		for _, c := range w.Containers {
+			found := false
+			for _, cc := range compare.Containers {
+				if c.IsSame(cc) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+type DeviceStatus struct {
+	Connectivity  map[string]bool      `json:"connectivity"` //  hosts and whether this device can reach them or not
+	Microservices []MicroserviceStatus `json:"microservices"`
+	Workloads     []WorkloadStatus     `json:"workloads"`
+	LastUpdated   string               `json:"lastUpdated"`
+}
+
+func (w DeviceStatus) String() string {
+	return fmt.Sprintf(
+		"Connectivity: %v, "+
+			"Microservices: %v, "+
+			"Workloads: %v,"+
+			"LastUpdated: %v",
+		w.Connectivity, w.Microservices, w.Workloads, w.LastUpdated)
+}
+
+func (w DeviceStatus) IsSame(compare DeviceStatus) bool {
+	if !reflect.DeepEqual(w.Connectivity, compare.Connectivity) {
+		return false
+	} else if !MsArrayIsSame(w.Microservices, compare.Microservices) {
+		return false
+	} else if !WlArrayIsSame(w.Workloads, compare.Workloads) {
+		return false
+	}
+
+	return true
+}
+
+// compare two microservice status arrays
+func MsArrayIsSame(a1 []MicroserviceStatus, a2 []MicroserviceStatus) bool {
+	if len(a1) != len(a2) {
+		return false
+	} else {
+		for _, item1 := range a1 {
+			found := false
+			for _, item2 := range a2 {
+				if item1.IsSame(item2) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// compare two workload array
+func WlArrayIsSame(a1 []WorkloadStatus, a2 []WorkloadStatus) bool {
+	if len(a1) != len(a2) {
+		return false
+	} else {
+		for _, item1 := range a1 {
+			found := false
+			for _, item2 := range a2 {
+				if item1.IsSame(item2) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Report the containers status and connectivity status to the exchange.
+func (w *GovernanceWorker) reportStatus() {
+
+	go func() {
+
+		// This is the amount of time for the governance routine to wait.
+		waitTime := w.Config.Edge.ReportDeviceStatusInterval
+		if waitTime == 0 { // never report status when the interval is set to 0
+			glog.V(5).Info("ReportDeviceStatusInterval is 0. The status report to the exchange is turned off.")
```

Comment:
This should be an Infof without the verbose setting, just glog.Infof(... because we always want this message to come out in the log.

Reply:
Done